### PR TITLE
Changes keep effect and planning code

### PR DESCRIPTION
**Pull request by NishanthJKumar**
This PR introduces two main changes:

1. pre-image backchaining at planning time: instead of running the `expected_atoms` check on the atoms sequence yielded by planning, we use a simple backchaining strategy to compute a 'necessary' atoms sequence by computing the necessary image at every timestep in the plan. This is exactly the number of atoms that need be true at a certain timestep for the rest of the plan to succeed (consequently, this is the smallest set on which we can run the `expected_atoms` check).
2. no ignoring add-effects at learning time: previously, we would make ignore effects from any unmodeled add or delete effects. However, because of positive preconditions, we don't need to do this for add effects. So we now just make ignore effects from unmodeled delete effects. As a corollary to this change, we modify the keep effects procedure to not just check ignore effects but also delete effects.

```
Git commit hashes seen in results/:
311624a54a7acdf26cada1c316f16e443a8938d9


AGGREGATED DATA OVER SEEDS:
                                                  NUM_TRAIN_TASKS     NUM_SOLVED AVG_NUM_PREDS AVG_TEST_TIME   AVG_NODES_CREATED LEARNING_TIME  NUM_SEEDS
EXPERIMENT_ID                                                                                                                                            
painting_backchaining_10demo                         10.00 (0.00)   45.50 (4.46)  15.00 (0.01)   0.46 (0.17)    2050.96 (833.18)  54.67 (3.18)         10
painting_backchaining_25demo                         25.00 (0.00)   49.40 (0.80)  15.00 (0.01)   0.39 (0.07)    2369.40 (825.36)  53.99 (2.03)         10
painting_backchaining_50demo                         50.00 (0.00)   49.50 (0.67)  15.00 (0.01)   0.43 (0.12)   2746.11 (1087.58)  58.42 (4.76)         10
painting_backchaining_5demo                           5.00 (0.00)   13.60 (9.48)  14.96 (0.08)   1.56 (1.17)   1373.40 (1080.24)  50.35 (1.99)         10
painting_oracle                                       0.00 (0.00)   50.00 (0.00)  15.00 (0.01)   0.38 (0.09)    2301.79 (738.99)   0.00 (0.00)         10
repeated_nextto_painting_backchaining_10demo         10.00 (0.00)   37.40 (7.19)  19.00 (0.01)   0.99 (0.27)   2966.33 (2126.80)  84.73 (4.18)         10
repeated_nextto_painting_backchaining_25demo         25.00 (0.00)   45.70 (3.13)  19.00 (0.00)   0.70 (0.05)    1748.50 (462.34)  90.73 (3.28)         10
repeated_nextto_painting_backchaining_50demo         50.00 (0.00)   49.50 (0.67)  19.00 (0.01)   0.75 (0.07)    1833.85 (418.57)  97.74 (3.37)         10
repeated_nextto_painting_backchaining_5demo           5.00 (0.00)  23.60 (10.89)  18.97 (0.09)   1.10 (0.42)   4408.00 (3408.59)  76.88 (5.46)         10
repeated_nextto_painting_oracle                       0.00 (0.00)   49.30 (1.00)  19.00 (0.01)   1.33 (0.17)   6597.14 (2088.21)   0.00 (0.00)         10
repeated_nextto_single_option_backchaining_10demo    10.00 (0.00)   50.00 (0.00)   3.00 (0.00)   0.04 (0.01)        88.50 (2.35)  19.19 (2.36)         10
repeated_nextto_single_option_backchaining_25demo    25.00 (0.00)   50.00 (0.00)   3.00 (0.00)   0.03 (0.00)        88.50 (2.35)  19.50 (1.59)         10
repeated_nextto_single_option_backchaining_50demo    50.00 (0.00)   50.00 (0.00)   3.00 (0.00)   0.03 (0.00)        88.50 (2.35)  18.01 (1.22)         10
repeated_nextto_single_option_backchaining_5demo      5.00 (0.00)   49.90 (0.30)   3.00 (0.00)   0.04 (0.01)       83.84 (14.86)  21.08 (3.21)         10
repeated_nextto_single_option_oracle                  0.00 (0.00)   50.00 (0.00)   3.00 (0.00)   0.02 (0.00)        88.50 (2.35)   0.00 (0.00)         10
satellites_backchaining_10demo                       10.00 (0.00)   22.90 (7.75)  13.00 (0.00)   1.79 (0.54)   6022.24 (3347.84)  10.89 (0.98)         10
satellites_backchaining_25demo                       25.00 (0.00)   38.00 (5.39)  13.00 (0.00)   1.30 (0.23)   5035.53 (2361.28)  12.19 (1.08)         10
satellites_backchaining_50demo                       50.00 (0.00)   47.20 (1.40)  13.00 (0.00)   1.10 (0.21)   6302.40 (2149.91)  13.88 (0.63)         10
satellites_backchaining_5demo                         5.00 (0.00)    7.30 (9.66)  13.00 (0.00)   1.77 (1.22)  7436.84 (10126.92)  10.96 (1.26)         10
satellites_oracle                                     0.00 (0.00)   49.10 (1.04)  13.00 (0.00)   0.83 (0.14)   5317.83 (1774.52)   0.00 (0.00)         10
satellites_simple_backchaining_10demo                10.00 (0.00)   33.30 (6.84)  13.00 (0.00)   0.91 (0.42)     393.45 (203.84)  10.28 (0.60)         10
satellites_simple_backchaining_25demo                25.00 (0.00)   44.80 (2.60)  13.00 (0.00)   0.35 (0.17)      287.88 (43.37)  11.68 (0.89)         10
satellites_simple_backchaining_50demo                50.00 (0.00)   49.10 (0.70)  13.00 (0.00)   0.16 (0.07)      280.04 (45.77)  12.74 (0.86)         10
satellites_simple_backchaining_5demo                  5.00 (0.00)    7.30 (8.12)  13.00 (0.00)   1.56 (2.07)     532.54 (343.75)  10.18 (0.97)         10
satellites_simple_oracle                              0.00 (0.00)   50.00 (0.00)  13.00 (0.00)   0.04 (0.00)      280.87 (37.98)   0.00 (0.00)         10
screws_backchaining_10demo                           10.00 (0.00)   50.00 (0.00)   4.00 (0.00)   0.10 (0.00)       117.32 (1.06)   0.17 (0.01)         10
screws_backchaining_25demo                           25.00 (0.00)   50.00 (0.00)   4.00 (0.00)   0.10 (0.00)       117.32 (1.06)   0.42 (0.02)         10
screws_backchaining_50demo                           50.00 (0.00)   50.00 (0.00)   4.00 (0.00)   0.10 (0.00)       117.32 (1.06)   0.86 (0.03)         10
screws_backchaining_5demo                             5.00 (0.00)   50.00 (0.00)   4.00 (0.00)   0.10 (0.00)       117.32 (1.06)   0.10 (0.00)         10
screws_oracle                                         0.00 (0.00)   50.00 (0.00)   4.00 (0.00)   0.07 (0.00)       116.32 (1.06)   0.00 (0.00)         10
```

### CODE REVIEW

**Review by tomsilver — COMMENTED** [review]
looking very good! one high level question: do we need to change `task_plan_with_option_plan_constraint`? (why or why not?)

File: tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py [hunk 1]
```diff
@@ -805,7 +805,7 @@ def test_keep_effect_data_partitioning():
     Preconditions: [MachineConfigurableWhileOff(?x0:machine_type)]
```
Comment (tomsilver) [inline tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py:805]: it's nice to see a number of these operators getting simpler, and planning still working!

**Comment by NishanthJKumar** [comment]
great question: I originally thought the answer was 'no', but I now think it is 'yes'. Specifically, we need to modify `task_plan_with_option_plan_constraint` to not check the expected atoms after every action execution, but rather compute an entire task plan, then find the necessary atoms sequence, then check that for every timestep, the necessary atoms are a subset of the actually-observed atoms.

I will think about the easiest way to make this change and try implementing sometime later today/tomorrow!

**Comment by NishanthJKumar** [comment]
Okay - so I'm actually pretty stuck on how to do this. Let's chat about it during our meeting tomorrow @tomsilver ?

**Comment by tomsilver** [comment]
Yeah it looked tricky to me too, chatting sounds good

**Comment by NishanthJKumar** [comment]
@tomsilver change made!

File: predicators/planning.py [hunk 2]
```diff
@@ -666,10 +666,10 @@ def _get_successor_with_correct_option(
                 continue
             if applicable_nsrt.option_objs != gt_objects:
                 continue
-            next_atoms = utils.apply_operator(applicable_nsrt, set(atoms))
-            if atoms_seq is not None and \
-                not next_atoms.issubset(expected_next_atoms):
+            if atoms_seq is not None and not applicable_nsrt.preconditions\
```
Comment (tomsilver) [inline predicators/planning.py:669]: this looks pretty ugly, can we split the new line in any other way?

**Review by tomsilver — APPROVED** [review]
lgtm! just one cosmetic complaint

**Pull request merged**